File: list/list.py
from listNode import Node

class list:
    def __init__(self):
        self.head = None

    def createNode(self, val):
        node = Node()
        node.setVal(val)
        node.setNext(None)
        return node

    def getListLength(self):
        length = 0
        current = self.head
        while current is not None:
            length += 1
            current = current.getNext()
        return length

    def addNodeAtBegin(self, val):
        """
            Time:   O(1)
        """
        node = self.createNode(val)
        node.setNext(self.head)
        self.head = node

    def addNodeAtEnd(self, val):
        """
            Time:   O(n)
        """
        node = self.createNode(val)
        current = self.getHead()
        if current is None:
            self.head = node
        else:
            while current.getNext() is not None:
                current = current.getNext()
            current.setNext(node)

    def addNodeAtPosX(self, val, pos):
        """
            Time:   O(n)
        """
        node = self.createNode(val)
        current = self.getHead()
        # Assuming node map or may not have p elements
        if current is None:
            # Insert val at head
            self.head = node
        else:
            if (pos == 0) or (pos == 1):
                # Handling the case of insertion at first position separately
                node.setNext(current)
                self.head = node
                return
            i = 1
            # Insert val at position 'p' or at end of list, which ever is smaller
            while (current.getNext() is not None) and (i < pos-1):
                current = current.getNext()
                i += 1
            node.setNext(current.getNext())
            current.setNext(node)

    def deleteFromBegin(self):
        """
            Time:   O(1)
        """
        if self.head is None:
            return
        current = self.head
        self.head = current.getNext()
        del current

    def deleteFromEnd(self):
        """
            Time:   O(n)
        """
        if self.head is None:
            return
        current = self.head
        if current.getNext() is None:
            self.head = None
            del current
        else:
            while current.getNext().getNext() is not None:
                current = current.getNext()
            ele = current.getNext()
            del ele
            current.setNext(None)

    def deleteFromPosX(self, pos):
        if self.head is None:
            return
        prev = self.head
        current = prev.getNext()

        if (pos == 0) or (pos == 1):
            self.head = prev.getNext()
            del prev
        else:
            i = 1
            while (current is not None) and i < pos-1:
                prev = current
                current = current.getNext()
                i += 1
            if current is None:
                prev.setNext(current)
            else:
                prev.setNext(current.getNext())
            del current


    def printList(self):
        """
            Time:   O(n)
        """
        current = self.head
        while current is not None:
            print current.getVal(),'->',
            current = current.getNext()
        print 'NULL'

    def getHead(self):
        return self.head

    def setHead(self, head):
        self.head = head


if __name__ == '__main__':
    l = list()
    l.setHead(l.createNode(1))
    l.deleteFromBegin()
    l.printList()
    print 'Length of list is:',l.getListLength()
    l.addNodeAtBegin(2)
    l.addNodeAtBegin(3)
    l.printList()
    l.addNodeAtBegin(4)
    l.addNodeAtBegin(5)
    l.printList()
    l.addNodeAtEnd(10)
    l.addNodeAtEnd(15)
    l.printList()
    l.addNodeAtPosX(100,1)
    l.addNodeAtPosX(200,9)
    l.printList()

    print 'Length of list is:',l.getListLength()
    
    l.deleteFromBegin()
    l.printList()
    print 'Length of list is:',l.getListLength()
    l.deleteFromEnd()
    l.printList()
    print 'Length of list is:',l.getListLength()
    l.deleteFromPosX(6)
    l.printList()
    print 'Length of list is:',l.getListLength()

File: DataStructuresPython/list/List.java
package list;

import list.ListNode;

public class List {
	private ListNode head;

	public List(){
		this.head = null;
	}

	// add a node at the end of the list
	public void addNode (Integer val){
		ListNode node = new ListNode();
		node.setVal(val);
		if(this.head == null){
			this.head = node;
		}else{
			ListNode current = this.head;
			while(current.getNext() != null){
				current = current.getNext();
			}
			current.setNext(node);
		}
	}

	// delete the first node with the given value from the list
	public void deleteNode (Integer val){
		ListNode current = this.head;
		if (current == null){
			return;
		}else if(current.getVal() == val){
			this.head = current.getNext();
		}else{
			while (current.getNext() != null){
				if(current.getNext().getVal() == val){
					// This is the node to be deleted
					current.setNext(current.getNext().getNext());
				}
				current = current.getNext();
			}
		}
	}

	public void printList(){
		ListNode current = this.head;
		System.out.println("Printing out the list");
		while (current != null){
			System.out.print(current.getVal() + " --> ");
			current = current.getNext();
		}
		System.out.println("Null");
	}

	public ListNode getHead(){
		return this.head;
	}

	public void setHead (ListNode head){
		this.head = head;
	}

	public static void main (String[] args){
		List lst = new List();

		lst.deleteNode(30);
		lst.printList();

		lst.addNode(10);
		lst.printList();

		lst.addNode(20);
		lst.addNode(30);
		lst.addNode(40);
		lst.addNode(50);
		lst.printList();

		lst.deleteNode(30);

		lst.printList();
	}
}